lib/rucio/common/types.py

```diff
@@ -99,3 +99,43 @@ class InternalScope(InternalType):
     '''
     def __init__(self, scope, vo='def', fromExternal=True):
         super(InternalScope, self).__init__(value=scope, vo=vo, fromExternal=fromExternal)
+
+
+class RSEDomainLANDict(TypedDict):
+    read: int
+    write: int
+    delete: int
+
+
+class RSEDomainWANDict(TypedDict):
+    read: int
+    write: int
+    delete: int
+    third_party_copy_read: int
+    third_party_copy_write: int
+
+
+class RSEDomainsDict(TypedDict):
+    lan: dict[str, RSEDomainLANDict]
+    wan: dict[str, RSEDomainWANDict]
+
+
+class RSEProtocolDict(TypedDict):
+    hostname: str
+    scheme: str
+    port: int
+    prefix: str  # FIXME: typing.NotRequired
+    impl: str
+    domains: dict[str, RSEDomainsDict]
+    extended_attributes: dict[str, str]  # FIXME: typing.NotRequired
+
+
+class RSESettingsDict(TypedDict):
+    id: str
+    rse: str
+    type: str
+    volatile: bool
+    verify_checksum: bool
+    deterministic: bool
+    domain: str
+    protocols: list[dict[str, RSEProtocolDict]]
```

Comment:
What about the other keys set below?

https://github.com/rucio/rucio/blob/c4f4eb9902fe0be6ff52d22357f763769c8e9860/lib/rucio/core/rse.py#L1399-L1417

Reply:
Just figured it out. I didn't include these keys because they weren't obtained by `rse_settings` parameter but they are still returned by `_format_get_rse_protocols `. I think they are the ones that should be marked with ` # FIXME: typing.NotRequired`. Thank you for pointing it out.

Reply:
> I think they are the ones that should be marked with ` # FIXME: typing.NotRequired`.

Why? Is there a code path which could result in the structure not containing those keys? Consider the following example:

```python
from typing import NotRequired, TypedDict

class FooDict(TypedDict):
    foo: str
    bar: NotRequired[str]

def foo(add_bar: bool = True) -> FooDict:
    d = {'foo': 'foo'}
    if add_bar:
        d['bar'] = 'bar'
    return d
```

Here, it makes sense to define `bar` as `NotRequired` because a `foo(add_bar=False)` call will result in its absence from the returned dictionary. Is it the case for the ones in `RSESettingsDict`?

Reply:
Oh, if it should be used in such cases then I don't see where `NotRequired` could be possibly implemented:

- `rse_attributes` isn't always used as an argument for function but the default value `None` will be returned anyway and if-else statements will give values to `lfn2pfn_algorithm`, `verify_checksum` and `sign_url` in `_format_get_rse_protocols`.
- It seems that `get_rse_protocols` and `get_rse_info`  also do not change structure of returned dictionaries.
- Then `get_rse_info` passes `rse_settings` as an argument to parameters of `rsemanager` functions.

Could I missed something?

Reply:
But if `NotRequired` is not necessary, then that is a good thing! I mentioned it as a solution to a possible situation, as describe in my example above.